src/merge_maps_kinematic.cpp

```diff
@@ -44,61 +44,76 @@ void MergeMapTool::SetConfigs()
   }
+  for(std::vector<karto::Dataset*>::iterator dataset_it = dataset_vec_.begin(); dataset_it != dataset_vec_.end(); ++dataset_it )
+  {
+    delete *dataset_it;
+  }
 }
 
 /*****************************************************************************/
-bool MergeMapTool::AddSubmapCallback(slam_toolbox::AddSubmap::Request &req,
+bool MergeMapsKinematic::AddSubmapCallback(slam_toolbox::AddSubmap::Request &req,
                                      slam_toolbox::AddSubmap::Response &resp)
 /*****************************************************************************/
 {
   // find if file exists
-  const std::string filename = req.filename + std::string(".st");
-  if (!FileExists(filename))
+  const std::string filename = req.filename;
+  karto::Mapper* mapper = new karto::Mapper;
+  karto::Dataset* dataset = new karto::Dataset;
+  serialization::Read(filename, mapper, dataset);
+  karto::LaserRangeFinder* laser = dynamic_cast<karto::LaserRangeFinder*>(dataset->GetObjects()[0]);
+  if (lasers_.find(laser->GetName().GetName())==lasers_.end())
   {
-    ROS_ERROR("MergeMapTool: Failed to open requested submap %s.", filename.c_str());
-    return true;
+    lasers_[laser->GetName().GetName()] = laser;
+    dataset->Add(laser);
+    dataset_vec_.push_back(dataset);
   }
 
-  karto::Mapper* mapper = new karto::Mapper();
-  serialization::Read(filename, mapper);
-  karto::LocalizedRangeScanVector scans = mapper->GetAllProcessedScans(); // TODO should I be saving a vect or of these mapper objects? / scans, A: YES
-
-  // num_submaps_++ and name frame appropriately
+  karto::LocalizedRangeScanVector scans = mapper->GetAllProcessedScans();
+  scans_vec_.push_back(scans);
   num_submaps_++;
-  submap_locations_[num_submaps_] = Eigen::Vector3d(0.,0.,0.);
 
   // create and publish map with marker that will move the map around
   sstS_.push_back(nh_.advertise<nav_msgs::OccupancyGrid>( \
                             "/map_"+std::to_string(num_submaps_), 1, true));
   sstmS_.push_back(nh_.advertise<nav_msgs::MapMetaData>( \
                  "/map_metadata_" + std::to_string(num_submaps_), 1, true));
   ros::Duration(1.).sleep();
-
-  KartoToROSOccupancyGrid(scans);
-  map_.map.header.stamp = ros::Time::now();
-  map_.map.header.frame_id = "map_"+std::to_string(num_submaps_);
-  sstS_[num_submaps_].publish(map_.map);
-  sstmS_[num_submaps_].publish(map_.map.info);
+  nav_msgs::GetMap::Response map;
+  KartoToROSOccupancyGrid(scans,map);
+
+  tf::Transform transform;
+  transform.setOrigin(tf::Vector3(map.map.info.origin.position.x + map.map.info.width * map.map.info.resolution / 2.0,\
+                                  map.map.info.origin.position.y + map.map.info.height * map.map.info.resolution / 2.0, 0.));
+  map.map.info.origin.position.x = - (map.map.info.width * map.map.info.resolution / 2.0);
+  map.map.info.origin.position.y = - (map.map.info.height * map.map.info.resolution / 2.0);
+  transform.setRotation(tf::createQuaternionFromRPY(0,0,0));
```

Comment:
remove this line. Instead when you make the transform, use setIdentity(), then set the origin and delete this. You're doing quaternion math equivalent to (0,0,0,1)